integrations/azure_ai_search/src/haystack_integrations/document_stores/azure_ai_search/document_store.py

```diff
@@ -0,0 +1,414 @@
+        :param embedding_dimension: Dimension of the embeddings.
+        :param metadata_fields: A dictionary of metatada keys and their types to create
+        additional fields in index schema. As fields in Azure SearchIndex cannot be dynamic, it is necessary to specify the metadata fields in advance.
+        :param vector_search_configuration: Configuration option related to vector search.
+        Default configuration uses the HNSW algorithm with cosine similarity to handle vector searches.
+
+        :param kwargs: Optional keyword parameters for Azure AI Search.
+        Some of the supported parameters:
+            - `api_version`: The Search API version to use for requests.
+            - `audience`: sets the Audience to use for authentication with Azure Active Directory (AAD).
+            The audience is not considered when using a shared key. If audience is not provided, the public cloud audience will be assumed.
+
+        For more information on parameters, see the [official Azure AI Search documentation](https://learn.microsoft.com/en-us/azure/search/)
+        """
+
+        azure_endpoint = azure_endpoint or os.environ.get("AZURE_SEARCH_SERVICE_ENDPOINT")
+        if not azure_endpoint:
+            msg = "Please provide an Azure endpoint or set the environment variable AZURE_OPENAI_ENDPOINT."
+            raise ValueError(msg)
+        api_key = api_key or os.environ.get("AZURE_SEARCH_API_KEY")
+
+        self._client = None
+        self._index_client = None
+        self._index_fields = None  # stores all fields in the final schema of index
+        self._api_key = api_key
+        self._azure_endpoint = azure_endpoint
+        self._index_name = index_name
+        self._embedding_dimension = embedding_dimension
+        self._dummy_vector = [-10.0] * self._embedding_dimension
+        self._metadata_fields = metadata_fields
+        self._vector_search_configuration = vector_search_configuration or DEFAULT_VECTOR_SEARCH
+        self._create_index = create_index
+        self._kwargs = kwargs
+
+    @property
+    def client(self) -> SearchClient:
+
+        if isinstance(self._azure_endpoint, Secret):
+            self._azure_endpoint = self._azure_endpoint.resolve_value()
+
+        if isinstance(self._api_key, Secret):
+            self._api_key = self._api_key.resolve_value()
+        credential = AzureKeyCredential(self._api_key) if self._api_key else DefaultAzureCredential()
+        try:
+            if not self._index_client:
+                self._index_client = SearchIndexClient(self._azure_endpoint, credential, **self._kwargs)
+            if not self.index_exists(self._index_name):
+                # Create a new index if it does not exist
+                logger.debug(
+                    "The index '%s' does not exist. A new index will be created.",
+                    self._index_name,
+                )
+                self.create_index(self._index_name)
```

Comment:
I noticed that the `create_index` function signature allows kwargs to be passed in. Should they be passed in here? I was thinking about the case where the user wants to enable semantic re-ranking/search in their index, which can be toggled in the `SearchIndex()` call below.

Reply:
Spoke with my team, and we want to prioritize allowing users to toggle on semantic re-ranking. This is a setting should be set at index creation time, so adding a toggle for this setting in the `AzureAISearchDocumentStore` seems like the way forward. At retrieval time, we can allow the user to toggle semantic ranking, which will only work if their index was initialized with it. It is also worth noting that on the free tier of the semantic ranker, the user has a quota of 1,000 semantic requests per month, and then they can upgrade beyond that.

Thoughts on implementing support for this?

Reply:
That's a great suggestion. I believe we can add support for semantic re-ranking during index creation. We usually allow passing `kwargs` specific to the database, but I'll need to check the documentation to understand how this will affect index creation and the remaining methods. Regardless, I'll add support for semantic re-ranking after referring to API.

Reply:
Decided to introduce this change as a separate PR to avoid extending the current PR further.